### PR TITLE
[REEF-1380] HeartbeatManager can trigger a NullReferenceException if …

### PR DESCRIPTION
**Pull request by afchung**
…Evaluator operation state is [OPERATIONAL] and runtime state is [FAILED]

This addressed the issue by
  * Checking for DONE states before sending heartbeat.
  * Only accessing _driverConnection when operation state is RECOVERY.
  * Using Lazy to instantiate _driverConnection.

JIRA:
  [REEF-1380](https://issues.apache.org/jira/browse/REEF-1380)

### CODE REVIEW

**Comment by afchung** [comment]
I have tested Driver Restart on an HDInsight cluster.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs [hunk 1]
```diff
@@ -90,6 +93,24 @@ internal sealed class HeartBeatManager : IHeartBeatManager
                 _clock = settings.RuntimeClock;
                 _heartBeatPeriodInMillSeconds = settings.HeartBeatPeriodInMs;
                 _maxHeartbeatRetries = settings.MaxHeartbeatRetries;
+                _driverConnection = new Lazy<IDriverConnection>(
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs]: This looks a lot like a nicer approach to `InjectionFuture`? Or should we use `InjectionFuture` here?

**Comment by markusweimer** [comment]
LGTM with one question

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs (hunk 1 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs]: Sure, that works.

**Comment by afchung** [comment]
@markusweimer I've addressed your comment and tested on HDInsight, please have another look. Thanks!

**Comment by markusweimer** [comment]
LGTM, will test and merge

**Pull request closed without merging**